### ISSUE TITLE
Accepting sockets with long external poll 

### ISSUE THREAD

**Issue by ralight** [warmcat/libwebsockets#2057]
What is the correct way to handle waiting on the listening server sockets that call accept() when using the external poll interface? As far as I can tell, `LWS_CALLBACK_ADD_POLL_FD` only operate on the client sockets.

If you compile lws with ext poll enabled, then run `libwebsockets-server-extpoll`, and `libwebsockets-test-client`, you can see the client connects quite happily.

If you now modify test-server.c and change:
```
n = poll(pollfds, count_pollfds, 50);
```
to
```
n = poll(pollfds, count_pollfds, 5000);
```
then recompile, restart, and try to reconnect `libwebsockets-test-client`, you will see that it fails as it is unable to complete the connect handshake in a reasonable time. This means it cannot be waiting on the listening socket.

Is there a way to access the listening sockets, so they can be added to `pollfds` in this case where we want a long timeout?

**Comment by lws-team** [comment]
This is caused by a patch in April 2018 modernizing the test apps to use the plugins.  It removed most of the external poll special-casing.

I pushed a patch on master + v4.1-stable so it can work again... it also adds the ripe sul handling inside the existing public api, it's worth referring to what that looks like

https://libwebsockets.org/git/libwebsockets/tree/test-apps/test-server.c#n662

External Poll is just a legacy hack at this point... there are supported ways to join all the other loops like libuv, libev, libevent and glib because their libs have formal interfaces.  But for rolled-your-own poll in external apps, there's no standardized way to integrate with it.  It's not going to be around forever and the cleanest way with a future is update your app to use an event lib instead of poll, and have lws join that using the foreign event loop support.

**Comment by ralight** [comment]
You may view the external poll support as a hack, and in the background it may be, but for the developer it is a very simple and interface to use. Saying "rewrite your app to use an event loop" is pretty ill considered of your audience. Just because nobody has informed you that the test server wasn't working doesn't mean there aren't people adding that functionality into their code, and that's not to mention all of the existing applications that rely on what you are planning on removing.

The event libraries all have their problems. The only one that I would consider worth migrating to is libuv, but the external loop support, which lws uses, is explicitly marked as "do not use if you need good performance", so it would be a downgrade from my current situation. It also requires a lot of hackery to make it work with TLS.

The changes you've made affect the poll for the client sockets only, and the maximum timeout seems to be forced at 1000 milliseconds for some reason, even on an idle server. What I would like is a function that tells me when lws needs to next carry out an event, but with a very high default value if there are genuinely no events waiting.

I would also like access to the listening sockets so I can make a single call to epoll per iteration of my loop, and hence be able to have full control over the timeout assigned to that call, and the latency that goes along with it. At the moment, I could quite happily call epoll_wait(..., ..., 8640000) on an idle server and everything would work fine, with the sole exception of waiting on lws listening sockets, which have a separate system call. If I could get the listening sockets, and a function to call when a client connected to that socket, then I could integrate everything into the single epoll_wait call and hence remove at least one unnecessary system call and get completely deterministic timeouts.

**Comment by lws-team** [comment]
> You may view the external poll support as a hack

Yeah it's a mistake on may part: as the author you should probably take my view about that into closer consideration.

> but for the developer it is a very simple and interface to use.

No.... it retcons callbacks about fd wait management (and locking...) on to what was originally a ws connection handling callback, and a specific one at that.  That's a strange historical thing nobody would choose to do from a clean slate.  It's "very simple" from your perspective because you have implemented it, and keeping it means you don't have to change your code.

The whole business of trying to glue different stacks together around a rolled-your-own poll loop is not the way to do it and leads to hacks like that.  In contrast event lib interfaces are clean and generic.

I have contained EXTERNAL_POLL with a cmake option that is default-off for a while now, it's not causing either bloat or immediate problems that it should be ripped out today, but eventually it will formally deprecated and then removed.

> libuv... "do not use if you need good performance"

Sorry where is this?  I grepped for it in lws but not hits.  If I wrote it I'm not sure any more what it was apropos of, since one advantage of the formalized event loops is they have swappable backends for things like epoll(), which may give better performance.  lwsws that runs all my web services is based on lws libuv support for many years, it's very performant.

Having worked with uv you should be aware it's the only lib of the ones I support with async deferred close that makes it a bit harder to work with (for example, lws must refcount all handles and stick around to unwind them for foreign case).

> The changes you've made affect the poll for the client sockets only, and the maximum timeout seems to be forced at 1000 milliseconds for some reason, even on an idle server. What I would like is a function that tells me when lws needs to next carry out an event, but with a very high default value if there are genuinely no events waiting.

If you use either lws poll / select default loop, or an event lib, that's exactly how it works today... for eg, libuv, it sets a libuv timer to wake it at the next scheduled event before returning to the event loop.  For poll / select, it restricts your preferred wait timeout (0 means 2 weeks, which fits comfortably inside 31 bits) to the next scheduled event and sets the poll or select timeout accordingly.

There's no 1s limit... it used to be you had to run background processing every second for timeouts, but those all migrated to individual lws_sul at the same time and the 1Hz requirement went away too.  If it's truly idle, it should stay in the wait for up to 2 weeks.  The link I posted yesterday shows that call to the api that does that; the lws default poll loop for unix looks like this

https://libwebsockets.org/git/libwebsockets/tree/lib/plat/unix/unix-service.c#n65

> I would also like access to the listening sockets

The idea of lws is to take care of that, and the accepts, and that's what the test app does now.

If you use an event lib, lws will cooperate with it to place the listen socket watcher on the same epoll() loop without you having to do anything at all.  And if you're building on some other platform than linux that doesn't have epoll(), if supported by the event lib it will implement the same capability automagically without you having to grub around at that level in WSA or whatever.
